subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/internal/toolchains/DefaultToolChainSelector.java

```diff
@@ -39,28 +40,48 @@ public DefaultToolChainSelector(ModelRegistry modelRegistry) {
 
     @Override
     public <T extends NativePlatform> Result<T> select(Class<T> platformType) {
-        DefaultNativePlatform targetPlatform = new DefaultNativePlatform("current");
-        NativeToolChainInternal toolChain = (NativeToolChainInternal) modelRegistry.realize("toolChains", NativeToolChainRegistryInternal.class).getForPlatform(targetPlatform);
-        PlatformToolProvider toolProvider = toolChain.select(targetPlatform);
+        DefaultNativePlatform platformRequest = new DefaultNativePlatform("current");
 
-        T t = null;
+        NativeToolChainInternal toolChain = (NativeToolChainInternal) modelRegistry.realize("toolChains", NativeToolChainRegistryInternal.class).getForPlatform(platformRequest);
+        PlatformToolProvider toolProvider = toolChain.select(platformRequest);
+
+        T targetPlatform = null;
         if (CppPlatform.class.isAssignableFrom(platformType)) {
-            t = platformType.cast(new DefaultCppPlatform("current"));
+            targetPlatform = platformType.cast(new DefaultCppPlatform("current"));
         } else if (SwiftPlatform.class.isAssignableFrom(platformType)) {
-            t = platformType.cast(new DefaultSwiftPlatform("current"));
+            if (toolProvider.isAvailable()) {
+                try {
+                    targetPlatform = platformType.cast(new DefaultSwiftPlatform("current", SwiftVersion.of(toolProvider.getCompilerMetadata().getVersion())));
+                } catch (IllegalArgumentException ex) {
+                    targetPlatform = platformType.cast(new UnknownSwiftPlatform("current"));
+                }
+            } else {
+                targetPlatform = platformType.cast(new UnknownSwiftPlatform("current"));
+            }
+        }
+        return new DefaultResult<T>(toolChain, toolProvider, targetPlatform);
+    }
+
+    class UnknownSwiftPlatform extends DefaultNativePlatform implements SwiftPlatform {
+        public UnknownSwiftPlatform(String name) {
+            super(name);
+        }
+
+        @Override
+        public SwiftVersion getSwiftVersion() {
+            throw new IllegalStateException("Unknown Swift platform doesn't have a Swift version.");
```

Comment:
It's not really the version that isn't known, it's the whole platform. You shouldn't be able to query anything about the platform, or shouldn't be able to get the platform itself. Or we should bust these apart (if the version really represents the Swift source language version).